test/provisioner/provision_ibmcloud.go

```diff
@@ -5,33 +5,455 @@ package provisioner
+		ResourceGroup: &vpcv1.ResourceGroupIdentity{
+			ID: &IBMCloudProps.ResourceGroupID,
+		},
+		Name:          &[]string{IBMCloudProps.VpcName}[0],
+		ClassicAccess: &classicAccess,
+		//AddressPrefixManagement: &manual,
+	}
+	ibmcloudTrace(fmt.Sprintf("Creating VPC %s in ResourceGroupID %s.\n", IBMCloudProps.VpcName, IBMCloudProps.ResourceGroupID))
+	vpcInstance, _, err := IBMCloudProps.VPC.CreateVPC(options)
+	if err != nil {
+		return err
+	}
+
+	IBMCloudProps.VpcID = *vpcInstance.ID
+	ibmcloudTrace(fmt.Sprintf("Created VPC with ID %s in ResourceGroupID %s.\n", IBMCloudProps.VpcID, IBMCloudProps.ResourceGroupID))
+
+	if len(IBMCloudProps.VpcID) <= 0 {
+		return errors.New("VpcID is empty, unknown error happened when create VPC.")
+	}
+
+	sgoptions := &vpcv1.GetVPCDefaultSecurityGroupOptions{}
+	sgoptions.SetID(IBMCloudProps.VpcID)
+	defaultSG, _, err := IBMCloudProps.VPC.GetVPCDefaultSecurityGroup(sgoptions)
+	if err != nil {
+		return err
+	}
+
+	IBMCloudProps.SecurityGroupID = *defaultSG.ID
+	ibmcloudTrace(fmt.Sprintf("Got VPC default SecurityGroupID %s.\n", IBMCloudProps.SecurityGroupID))
+
+	return nil
+}
+
+func deleteVPC() error {
+	if !IBMCloudProps.IsProvNewVPC {
+		fmt.Printf("Do not delete because using existing VPC: %s\n", IBMCloudProps.VpcID)
+		return nil
+	}
+
+	deleteVpcOptions := &vpcv1.DeleteVPCOptions{}
+	deleteVpcOptions.SetID(IBMCloudProps.VpcID)
+	ibmcloudTrace(fmt.Sprintf("Deleting VPC with ID %s.\n", IBMCloudProps.VpcID))
+	_, err := IBMCloudProps.VPC.DeleteVPC(deleteVpcOptions)
+
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace(fmt.Sprintf("Deleted VPC with ID %s.\n", IBMCloudProps.VpcID))
+	return nil
+}
+
+func createSubnet() error {
+	if !IBMCloudProps.IsProvNewSubnet {
+		fmt.Printf("Using existing Subnet: %s\n", IBMCloudProps.SubnetID)
+		return nil
+	}
+
+	cidrBlock := getCidrBlock(IBMCloudProps.Region, IBMCloudProps.Zone)
+	if cidrBlock == "" {
+		return errors.New("Can not calculate cidrBlock from Region and Zone.")
+	}
+	options := &vpcv1.CreateSubnetOptions{}
+	options.SetSubnetPrototype(&vpcv1.SubnetPrototype{
+		ResourceGroup: &vpcv1.ResourceGroupIdentity{
+			ID: &IBMCloudProps.ResourceGroupID,
+		},
+		Ipv4CIDRBlock: &cidrBlock,
+		Name:          &[]string{IBMCloudProps.SubnetName}[0],
+		VPC: &vpcv1.VPCIdentity{
+			ID: &IBMCloudProps.VpcID,
+		},
+		Zone: &vpcv1.ZoneIdentity{
+			Name: &IBMCloudProps.Zone,
+		},
+	})
+	ibmcloudTrace(fmt.Sprintf("Creating subnet %s in VPC %s in Zone %s.\n", IBMCloudProps.SubnetName, IBMCloudProps.VpcID, IBMCloudProps.Zone))
+	subnet, _, err := IBMCloudProps.VPC.CreateSubnet(options)
+	if err != nil {
+		return err
+	}
+	IBMCloudProps.SubnetID = *subnet.ID
+	ibmcloudTrace(fmt.Sprintf("Created subnet with ID %s.\n", IBMCloudProps.SubnetID))
+
+	if len(IBMCloudProps.SubnetID) <= 0 {
+		return errors.New("SubnetID is empty, unknown error happened when create Subnet.")
+	}
+
+	return nil
+}
+
+func deleteSubnet() error {
+	if !IBMCloudProps.IsProvNewSubnet {
```

Comment:
I'm not sure how this can ever work? If you specify a subnet name/id then it errors with `Do not delete because using existing Subnet: 02f7-3592fe1c-de60-4922-abc8-924f4c29550d`, but if your don't then the delete can't work as it doesn't know what to do:
```
Deleting subnet with ID .

deleteSubnetOptions failed validation:
Key: 'DeleteSubnetOptions.ID' Error:Field validation for 'ID' failed on the 'ne' tag
```

Reply:
I'll comments out `deprovision` as it looks conflict with the e2e provisioner. The design is when cluster/vpc/subnet name provided, we'll create them and then delete them after e2e test. While if only vpc/subnet ID provided, it means the e2e test is using exiting one, we should not delete them after the test. Looks the algorithm does not align with the cluster-provisioner CLI. Will think about it and add `deprovision` in a new PR.

Reply:
Use vpc and subnet name only should be a better approach, find it firstly and then create it if it's not existing. Will address this in a new PR.